### PR TITLE
Add tests based on ctest

### PR DESCRIPTION
**Pull request by tmiasko**
Use ctest crate for more comprehensive ABI checks.

It would be also nice if ctest could auto-detect our incomplete types.
Existing users of ctest crate use enum {}, which works out of the box,
but that would be UB if dereferenced. On the other hand, those types
shouldn't be dereferenced either way.

Discovered problems:
* `GIOFuncs::io_read` - const mismatch
* `GTestLogMsg::nums` - long instead of long double

### CODE REVIEW

**Comment by tmiasko** [comment]
@sdroege, re offsets checks from https://github.com/gtk-rs/gir/pull/675

**Comment by sdroege** [comment]
The existing users are wrong then and should re-read the nomicon :)

How does it map the C type names to the Rust names exactly, and the field names?

Looks good otherwise, just sad that it needs to be a separate crate. Wouldbee nicer to have it integrated with the existing tests :)

**Comment by tmiasko** [comment]
> How does it map the C type names to the Rust names exactly, and the field names?

Generally uses names from Rust as is, but they can be renamed as necessary, for
example closure passed to `cfg.field_name`, renames field names that had to be
changd due to conflict with Rust keywords.

> Looks good otherwise, just sad that it needs to be a separate crate. Wouldbee nicer to have it integrated with the existing tests :)

I would rather drop existing tests, since everything they are doing should be
covered here as well, at least once constant tests are enabled.

**Comment by sdroege** [comment]
Ah so it's testing constant values (incl strings?), type sizes/alignment and field offsets? Anything else? :)

Nice, I'd prefer dropping the old tests too then

**Comment by tmiasko** [comment]
It can validate size, alignment, offsets, types, function signatures, and
constants (I didn't look what exactly constant checks do).

Constants and function signatures are disabled right now. Function signature
checks have to match C functions exactly, including const, volatile etc, so we
should wait at for next gir update, since there should be significant
improvement in those areas. Otherwise too many things are just incorrect.

**Comment by sdroege** [comment]
Checking function signatures is something I wanted since a long time, so that's great. How does it fail right now, for example? :)

What's missing for enabling constants?

I'd be fine with merging this PR if we can also cover constants and just get rid of the other tests at the same time. It seems like a big improvement.

But it can't be made into a test that is inside the crate, it must be a separate crate? Or could we have it inside the same crate if the build.rs parts are hidden behind a `if cfg!(test)`?

**Comment by tmiasko** [comment]
I added tests for gio and gdk.

**Comment by EPashkin** [comment]
Tests very fast, seems it can be added to both CI

**Comment by sdroege** [comment]
Before this is merged, can we also get constant checks to have a strict superset of the tests that currently exist?

**Comment by tmiasko** [comment]
Should I open a new PR for sys regen, and then rebase this one on top of it?

**Comment by EPashkin** [comment]
In gtk 4 non windows structs: "GtkPlug", "GtkPlugClass", "GtkSocket", "GtkSocketClass".
I also get error about "Window" that declared in `manual.rs`, it used with "GtkPlug" and "GtkSocket" so seems it also non windows.

**Comment by EPashkin** [comment]
Note, `Windows` skipped with `skip_type`,
also small fix to remove warning about unused type:
```diff
+++ b/gtk-sys-test/src/main.rs
@@ -4,6 +4,7 @@ extern crate gtk_sys;
 extern crate libc;
 
 use gtk_sys::*;
+#[cfg(not(windows))]
 use gtk_sys::xlib::*;
 
 include!(concat!(env!("OUT_DIR"), "/all.rs"));
```

Big thanks, only 2 crates to go and maybe its time add it to CI too.

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  "atk-sys-test",
+  "gdk-pixbuf-sys-test",
+  "gdk-sys-test",
+  "gio-sys-test",
+  "glib-sys-test",
+  "gobject-sys-test",
+  "gtk-sys-test",
+  "pango-sys-test",
+  "pangocairo-sys-test"
+]
```
Comment (sdroege) [inline Cargo.toml]: Can this be integrated with `cargo test --all` somehow?

Comment (sdroege) [reply Cargo.toml]: Should be possible by having an empty `main.rs` and having the current `main.rs` inside the `tests` directory.

**Comment by EPashkin** [comment]
@sdroege `make check` not enough?
IMHO there no other test in sys
Or there some system that need `cargo test` ?

@tmiasko Maybe add `for`

```diff
+++ b/Makefile
@@ -8,6 +8,7 @@ GIR_FILES = $(shell ls $(CONFIGS) | xargs -n1 perl -ne '$$name = $$1 if /^librar
 LIBS = $(CONFIGS:conf/gir-%.toml=%-sys/src/lib.rs)
 TEST_C_FILES = $(CONFIGS:conf/gir-%.toml=%-sys/tests/*.c)
 TEST_RS_FILES = $(CONFIGS:conf/gir-%.toml=%-sys/tests/*.rs)
+TEST_CRATES = $(CONFIGS:conf/gir-%.toml=%-sys-test)
 
 libs : $(LIBS)
 
@@ -26,13 +27,7 @@ regen_check: $(GIR) $(GIR_FILES)
 	@$(MAKE) -f $(THIS_FILE) libs
 	git diff -R --exit-code
 
-check:
-	cargo run -p glib-sys-test
-	cargo run -p gobject-sys-test
-	cargo run -p gio-sys-test
-	cargo run -p pango-sys-test
-	cargo run -p pangocairo-sys-test
-	cargo run -p gdk-sys-test
-	cargo run -p gdk-pixbuf-sys-test
-	cargo run -p atk-sys-test
-	cargo run -p gtk-sys-test
+check: 
+	for dir in $(TEST_CRATES); do \
+		cargo run -p $$dir; \
+	done
```

**Comment by tmiasko** [comment]
I would leave out CI integration for now. I removed workspaces as well, since
they interact badly with other things we are doing on CI.

File: Cargo.toml (hunk 1 above)
Comment (tmiasko) [reply Cargo.toml]: I removed workspaces, so there is no benefit from this now.

**Comment by sdroege** [comment]
> @sdroege `make check` not enough?
> IMHO there no other test in sys
> Or there some system that need `cargo test` ?

It seems annoying to have two build systems in there if cargo can already do what we need.

> I would leave out CI integration for now. I removed workspaces as well, since

IMHO all the crates of -sys should be in a workspace `Cargo.toml` at the top-level so you can simply run `cargo build --all` for example, and `cargo test --all`. What exactly does that break?

**Comment by tmiasko** [comment]
Well, cargo becomes pretty confused when path overrides are used to replace
dependency with itself. Not that I am surprised ;-).

**Comment by tmiasko** [comment]
> Where does that happen? I mean, how would you replace a dependency with itself and why? :)

CI uses path overrides to compile examples, see changes made in travis and appveyor for details.

File: Makefile [hunk 2]
```diff
@@ -27,12 +28,6 @@ regen_check: $(GIR) $(GIR_FILES)
...
-	cd gobject-sys-test && cargo run
-	cd gio-sys-test && cargo run
-	cd pango-sys-test && cargo run
-	cd pangocairo-sys-test && cargo run
-	cd gdk-sys-test && cargo run
-	cd gdk-pixbuf-sys-test && cargo run
-	cd atk-sys-test && cargo run
-	cd gtk-sys-test && cargo run
+	for crate in $(TEST_CRATES); do \
+	  cargo run -p $$crate; \
```
[the remark this answers is not in the document]
Comment (tmiasko) [reply Makefile]: Sounds reasonable, feel free to change it, if you prefer that.

**Comment by tmiasko** [comment]
@sdroege, if you want to track those please open a new issue. This PR is about
adding ABI offsets tests, so that we can have confidence in recent bit field
generator change. It isn't about refactoring the whole ABI testing infrastructure.

Listing all tangentially related things that are nice to have is not helpful.

**Comment by sdroege** [comment]
I disagree with that. Having two overlapping testsuites is not useful maintenance-wise, and I wouldn't merge it in this state because of that. It will only cause things to bitrot.

If @GuillaumeGomez and @EPashkin are fine with maintaining two testsuites then feel free to ignore me.

**Comment by GuillaumeGomez** [comment]
I clearly lack time to be able to maintain two tests suites...

**Comment by tmiasko** [comment]
> I disagree with that. Having two overlapping testsuites is not useful maintenance-wise, and I wouldn't merge it in this state because of that. It will only cause things to bitrot.

Would removing the overlap avoid the bitrot? I can remove the layout checks from existing tests, though I don't think how that solves anything.

I prefer to reach the target of single test suite in step-wise manner, rather in one giant PR.

**Comment by EPashkin** [comment]
@sdroege I don't see problem having 2 testiness for short time, as currently it don't used on CI anyway

**Comment by sdroege** [comment]
> @sdroege I don't see problem having 2 testiness for short time, as currently it don't used on CI anyway

If it's only for a short time, sure. @tmiasko are you intending to continue working on finishing the replacement of the testsuite in the next days/few weeks?

**Comment by tmiasko** [comment]
It is low priority for me since I don't see any immediate value added.

**Comment by tmiasko** [comment]
Feel free to revert bit field changes until complete replacement is ready for ABI tests.

**Comment by tmiasko** [comment]
@GuillaumeGomez could you tell what do you see as a maintenance problem in a new tests, maybe this is easily addressed?

**Comment by GuillaumeGomez** [comment]
Simple: time. :) If you have a time machine, I'd like to rent it!

**Comment by tmiasko** [comment]
Well, but what exactly about them make them take your time?

**Comment by GuillaumeGomez** [comment]
Maintenance mainly and potential updates when the API will change too.

**Comment by tmiasko** [comment]
That still doesn't answer the question what could be changed about test to make
that easier.

My main maintenance worry about ctests was the need to update the list of
incomplete types manually. That isn't necessary any longer.  I think that
ctests they are maintenance improvement in the long run, since you can detect
ABI problems upfront, rather than spend time debugging the issues.

If you are not interested in ctests just say so, since I don't really
understand your motivation with holding this back at this point.

**Comment by GuillaumeGomez** [comment]
I agree on all your points. I'm just afraid that adding this much tests will just make the whole maintenance heavier. Just imagine if we have to update all those tests every time we have to make an API changes: it'll be quite a time consumer. If you think you'll remain around and maintain those tests yourself, then I have no issue merging them, but it's a risk for us. Your call. :)

**Pull request closed without merging**